### PR TITLE
[Go Program Gen] binary, unary, and initial conditional expression support

### PR DESCRIPTION
**Pull request by EvanBoyle**
Go doesn't support ternary expressions so we must convert:

```
true ? 1 : 0
```
into a temp variable and assignment within if/else:

```go
var tmp0 float64
if true {
tmp0 = 1
} else {
tmp0 = 0
}
```

Need more conditional test cases for things like assignment within object cons, and nesting, but this seems like enough to review for now.

### CODE REVIEW

**Comment by EvanBoyle** [comment]
Added support for nested ternaries and ternaries within object cons expressions.

File: pkg/codegen/go/gen_program_ternaries.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		t, _ := ta.allocateExpression(x.TrueResult)
+		f, _ := ta.allocateExpression(x.FalseResult)
+		temp = &ternaryTemp{
+			Name:         fmt.Sprintf("tmp%d", len(ta.temps)),
+			VariableType: x.Type(),
+			Condition:    cond,
+			TrueResult:   t,
+			FalseResult:  f,
+		}
+		ta.temps = append(ta.temps, temp)
```
Comment (pgavlin) [inline pkg/codegen/go/gen_program_ternaries.go:50]: This representation loses any trivia information associated with the `?` and `:` tokens from the original `ConditionalExpression`. I suggest that we define `ternaryTemp` as

```go
type ternaryTemp struct {
	Name         string
	Value *model.ConditionalExpression
 }
```

And then rewrite the input expression's `Condition`, `True`, and `False` fields:

```suggestion
	case *model.ConditionalExpression:
		x.Condition, _ = ta.allocateExpression(x.Condition)
		x.TrueResult, _ = ta.allocateExpression(x.TrueResult)
		x.FalseResult, _ = ta.allocateExpression(x.FalseResult)
		temp = &ternaryTemp{
			Name:         fmt.Sprintf("tmp%d", len(ta.temps)),
			Value:    x,
		}
		ta.temps = append(ta.temps, temp)
```

File: pkg/codegen/go/gen_program_expressions.go [hunk 2]
```diff
@@ -304,16 +365,28 @@ func (g *generator) argumentTypeName(expr model.Expression, destType model.Type,
...
-	switch destType.(type) {
+	switch destType := destType.(type) {
 	case *model.OpaqueType:
-		return destType.String()
+		switch destType.Name {
+		case "number":
+			return "float64"
+		default:
+			return destType.String()
+		}
```
Comment (pgavlin) [inline pkg/codegen/go/gen_program_expressions.go:386]:
```suggestion
		switch destType {
		case model.NumberType:
			return "float64"
		default:
			return destType.Name
		}
```

File: pkg/codegen/go/gen_program_expressions.go [hunk 3]
```diff
@@ -79,11 +80,47 @@ func (g *generator) GenAnonymousFunctionExpression(w io.Writer, expr *model.Anon
...
 
-// GenConditionalExpression generates code for a ConditionalExpression.
-func (g *generator) GenConditionalExpression(w io.Writer, expr *model.ConditionalExpression) {}
+	g.Fgenf(w, "%.[1]*[2]v %[3]v %.[1]*[4]o", precedence, expr.LeftOperand, opstr, expr.RightOperand)
+}
+
+func (g *generator) GenConditionalExpression(w io.Writer, expr *model.ConditionalExpression) {
+	// Ternary expressions are not supported in go so we need to allocate temp variables in the parent scope.
+	// This is handled by lower expression and rewriteTernaries
+	contract.Failf("BUG: this conditional expression needs to be lowered.")
```
Comment (pgavlin) [inline pkg/codegen/go/gen_program_expressions.go:122]: Nit: we don't usually say "BUG" like this :P

```suggestion
	contract.Failf("unlowered conditional expression @ %v". expr.SyntaxNode().Range())
```

File: pkg/codegen/go/gen_program_ternaries.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	temps []*ternaryTemp
+}
+
+func (ta *tempAllocator) allocateExpression(x model.Expression) (model.Expression, hcl.Diagnostics) {
+	var temp *ternaryTemp
+	switch x := x.(type) {
+	case *model.ConditionalExpression:
+		cond, _ := ta.allocateExpression(x.Condition)
+		t, _ := ta.allocateExpression(x.TrueResult)
+		f, _ := ta.allocateExpression(x.FalseResult)
```
Comment (pgavlin) [inline pkg/codegen/go/gen_program_ternaries.go:39]: These shouldn't be necessary--you should be able to do `x.Condition, _ = ta.allocateExpression(x.Condition)`. Is that not working?

File: pkg/codegen/go/gen_program_ternaries.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+func (tt *ternaryTemp) Traverse(traverser hcl.Traverser) (model.Traversable, hcl.Diagnostics) {
+	return tt.Type().Traverse(traverser)
+}
+
+func (tt *ternaryTemp) SyntaxNode() hclsyntax.Node {
+	return syntax.None
+}
+
+type tempAllocator struct {
```
Comment (pgavlin) [inline pkg/codegen/go/gen_program_ternaries.go:29]: Nit: technically, what we're doing is spilling, so I'd name this `tempSpiller`
```suggestion
type tempSpiller struct {
```

File: pkg/codegen/go/gen_program_ternaries.go (hunk 4 above)
Comment (EvanBoyle) [reply pkg/codegen/go/gen_program_ternaries.go:39]: ah yes, forgot to drop the `:` when I tried to make the change the first time.

File: pkg/codegen/go/gen_program_ternaries.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+func (tt *ternaryTemp) SyntaxNode() hclsyntax.Node {
+	return syntax.None
+}
+
+type tempSpiller struct {
+	temps []*ternaryTemp
+}
+
+func (ta *tempSpiller) allocateExpression(x model.Expression) (model.Expression, hcl.Diagnostics) {
```
Comment (pgavlin) [inline pkg/codegen/go/gen_program_ternaries.go:33]:
```suggestion
func (ta *tempSpiller) spillExpression(x model.Expression) (model.Expression, hcl.Diagnostics) {
```

**Pull request merged**